tests/test_component.py

```diff
@@ -113,16 +152,315 @@ def get_context_data(self, name, css_class="", title="", **attrs):
+                self.content = bytes(content, "utf-8")
+
+        class SimpleComponent(component.Component):
+            response_class = MyResponse
+            template: types.django_html = "HELLO"
+
+        rendered = SimpleComponent.render_to_response()
+        self.assertIsInstance(rendered, MyResponse)
+
+        self.assertHTMLEqual(
+            rendered.content.decode(),
+            "HELLO",
+        )
+
+    @parametrize_context_behavior([("django", False), ("isolated", True)])
+    def test_render_slot_as_func(self, context_behavior_data):
+        is_isolated = context_behavior_data
+
+        class SimpleComponent(component.Component):
+            template: types.django_html = """
+                {% load component_tags %}
+                {% slot "first" required %}
+                {% endslot %}
+            """
+
+            def get_context_data(self, the_arg, the_kwarg=None, **kwargs):
+                return {
+                    "the_arg": the_arg,
+                    "the_kwarg": the_kwarg,
+                    "kwargs": kwargs,
+                }
+
+        def first_slot(ctx: Context):
+            self.assertIsInstance(ctx, Context)
+            # NOTE: Since the slot has access to the Context object, it should behave
+            # the same way as it does in templates - when in "isolated" mode, then the
+            # slot fill has access only to the "root" context, but not to the data of
+            # get_context_data() of SimpleComponent.
+            if is_isolated:
+                self.assertEqual(ctx.get("the_arg"), None)
+                self.assertEqual(ctx.get("the_kwarg"), None)
+                self.assertEqual(ctx.get("kwargs"), None)
+                self.assertEqual(ctx.get("abc"), None)
+            else:
+                self.assertEqual(ctx["the_arg"], "1")
+                self.assertEqual(ctx["the_kwarg"], 3)
+                self.assertEqual(ctx["kwargs"], {})
+                self.assertEqual(ctx["abc"], "def")
+
+            return "FROM_INSIDE_FIRST_SLOT"
+
+        rendered = SimpleComponent.render(
+            context={"abc": "def"},
+            args=["1"],
+            kwargs={"the_kwarg": 3},
+            slots={"first": first_slot},
+        )
+        self.assertHTMLEqual(
+            rendered,
+            "FROM_INSIDE_FIRST_SLOT",
+        )
+
+    @parametrize_context_behavior(["django", "isolated"])
+    def test_render_raises_on_missing_slot(self):
+        class SimpleComponent(component.Component):
+            template: types.django_html = """
+                {% load component_tags %}
+                {% slot "first" required %}
+                {% endslot %}
+            """
+
+        with self.assertRaises(TemplateSyntaxError):
+            SimpleComponent.render()
+
+        SimpleComponent.render(
+            slots={"first": "FIRST_SLOT"},
+        )
+
+    @parametrize_context_behavior(["django", "isolated"])
+    def test_render_with_include(self):
```

Comment:
In my fork my implementation had issue when I used `Component.render()` and the template included an `{% include %}` tag. This is why I associated the template with the context [here](https://github.com/EmilStenstrom/django-components/pull/521/files#diff-35c84ebceae50f7149f03bbcaf2b4f1289bc2984bf7859ca521b5dbe7960c396R431) when context was internally created inside `render` method.